sunscreen_runtime/src/linked.rs

```diff
@@ -0,0 +1,343 @@
+    fn from(err: crate::Error) -> Self {
+        LinkedProofError::SunscreenRuntimeError(err)
+    }
+}
+
+impl From<ProofError> for LinkedProofError {
+    fn from(err: ProofError) -> Self {
+        LinkedProofError::LogproofProofError(err)
+    }
+}
+
+/// Generate a set of generators for a single party where some of the
+/// generators are shared with another proof system.
+fn new_single_party_with_shared_generators(
+    gens_capacity: usize,
+    shared_generators: &[RistrettoPoint],
+    insertion_point: usize,
+    right_side_allocated: bool,
+) -> BulletproofGens {
+    let mut label = [b'G', 0, 0, 0, 0];
+    let mut g = GeneratorsChain::new(&label)
+        .take(gens_capacity)
+        .collect::<Vec<RistrettoPoint>>();
+
+    label[0] = b'H';
+    let mut h = GeneratorsChain::new(&label)
+        .take(gens_capacity)
+        .collect::<Vec<RistrettoPoint>>();
+
+    let mut index = insertion_point;
+    let mut left_side = !right_side_allocated;
+
+    // Insert the shared generators. Note that the order of the shared
+    // generators is reversed because the inputs in the R1CS BP are reversed
+    // after compilation.
+    for gen in shared_generators.iter() {
+        if left_side {
+            g[index] = *gen;
+            left_side = false;
+            index -= 1;
+        } else {
+            h[index] = *gen;
+            left_side = true;
+        }
+    }
+
+    // We can unwrap safely because we know that the generators are generated properly.
+    BulletproofGens::new_from_generators(vec![g], vec![h]).unwrap()
+}
+
+impl<Q: Ring + CryptoHash + ModSwitch<ZqRistretto> + RingModulus<4> + Ord> LinkedProof<Q> {
+    /**
+     * This function verifies a linked proof between a short discrete log proof
+     * (SDLP) and a R1CS bulletproof. An example use case is proving an encryption
+     * is valid (by SDLP) and that the encrypted message has some property (by R1CS
+     * Bulletproof).
+     *
+     * See the main documentation for more information.
+     *
+     * Arguments:
+     *
+     * * `lattice_problem`: The lattice problem to prove
+     * * `shared_indices`: The indices of the shared values between the SDLP and the
+     *                     R1CS bulletproof
+     * * `program`: The compiled ZKP program to prove
+     * * `private_inputs`: The private inputs to the ZKP program, not including the
+     *                     shared values
+     * * `public_inputs`: The public inputs to the ZKP program
+     * * `constant_inputs`: The constant inputs to the ZKP program
+     */
+    pub fn create<I>(
+        lattice_problem: &LatticeProblem<Q>,
+        shared_indices: &[(usize, usize)],
+        program: &CompiledZkpProgram,
+        private_inputs: Vec<I>,
+        public_inputs: Vec<I>,
+        constant_inputs: Vec<I>,
+    ) -> Result<Self, sunscreen_zkp_backend::Error>
+    where
+        I: Into<ZkpProgramInput> + Clone,
+    {
+        let backend = BulletproofsBackend::new();
```

Comment:
These changes leak the bulletproof backend implementation without the feature flag enabled - not sure what we want to do about that.

Reply:
I am going to put this behind a feature flag anyway (or potentially just the bulletproofs feature flag) because otherwise the wasm build won't work